pkg/controller/elasticsearch/remotecluster/remoteca/controller.go

```diff
@@ -0,0 +1,474 @@
+	associatedEs *esv1.Elasticsearch,
+) (map[types.NamespacedName]struct{}, error) {
+	span, _ := apm.StartSpan(ctx, "get_expected_remote_ca", tracing.SpanTypeApp)
+	defer span.End()
+	expectedRemoteClusters := make(map[types.NamespacedName]struct{})
+
+	// Add remote clusters declared in the Spec
+	for _, remoteCluster := range associatedEs.Spec.RemoteClusters.K8sLocal {
+		if !remoteCluster.IsDefined() {
+			continue
+		}
+		esRef := remoteCluster.ElasticsearchRef.WithDefaultNamespace(associatedEs.Namespace)
+		expectedRemoteClusters[esRef.NamespacedName()] = struct{}{}
+	}
+
+	var list esv1.ElasticsearchList
+	if err := c.List(&list, &client.ListOptions{}); err != nil {
+		return nil, err
+	}
+
+	// Seek for Elasticsearch resources where this cluster is declared as a remote cluster
+	for _, es := range list.Items {
+		for _, remoteCluster := range es.Spec.RemoteClusters.K8sLocal {
+			if !remoteCluster.IsDefined() {
+				continue
+			}
+			esRef := remoteCluster.ElasticsearchRef.WithDefaultNamespace(es.Namespace)
+			if esRef.Namespace == associatedEs.Namespace &&
+				esRef.Name == associatedEs.Name {
+				expectedRemoteClusters[k8s.ExtractNamespacedName(&es)] = struct{}{}
+			}
+		}
+	}
+
+	return expectedRemoteClusters, nil
+}
+
+// getCurrentRemoteCertificateAuthorities returns all the remote cluster keys for which a remote ca exists
+func getCurrentRemoteCertificateAuthorities(
+	ctx context.Context,
+	c k8s.Client,
+	associatedEs types.NamespacedName,
+) (map[types.NamespacedName]struct{}, error) {
+	span, _ := apm.StartSpan(ctx, "get_current_remote_ca", tracing.SpanTypeApp)
+	defer span.End()
+
+	currentRemoteClusters := make(map[types.NamespacedName]struct{})
+
+	// Get the remoteCA in the current namespace
+	var remoteCAList corev1.SecretList
+	if err := c.List(
+		&remoteCAList,
+		client.InNamespace(associatedEs.Namespace),
+		GetRemoteCaMatchingLabel(associatedEs.Name),
+	); err != nil {
+		return nil, err
+	}
+	for _, remoteCA := range remoteCAList.Items {
+		if remoteCA.Labels == nil {
+			continue
+		}
+		remoteNs := remoteCA.Labels[RemoteClusterNamespaceLabelName]
+		remoteEs := remoteCA.Labels[RemoteClusterNameLabelName]
+		currentRemoteClusters[types.NamespacedName{
+			Namespace: remoteNs,
+			Name:      remoteEs,
+		}] = struct{}{}
+	}
+
+	// Get the remoteCA where this cluster is involved in other namespaces
+	if err := c.List(
+		&remoteCAList,
+		client.MatchingLabels(map[string]string{
+			common.TypeLabelName:            TypeLabelValue,
+			RemoteClusterNamespaceLabelName: associatedEs.Namespace,
+			RemoteClusterNameLabelName:      associatedEs.Name,
+		}),
+	); err != nil {
+		return nil, err
+	}
+	for _, remoteCA := range remoteCAList.Items {
+		if remoteCA.Labels == nil {
+			continue
+		}
+		remoteEs := remoteCA.Labels[label.ClusterNameLabelName]
+		currentRemoteClusters[types.NamespacedName{
+			Namespace: remoteCA.Namespace,
+			Name:      remoteEs,
+		}] = struct{}{}
+	}
+
+	return currentRemoteClusters, nil
```

Comment:
I'm a bit lost in what is remote vs. what is local in this function.
Might be worth splitting into 2 different functions with more comments.